src/wmtk/TriMesh.cpp

```diff
@@ -117,11 +117,69 @@ std::optional<TriMesh::Tuple> TriMesh::Tuple::switch_face(const TriMesh& m) cons
     return loc;
 }
 
-bool TriMesh::collapse_edge(const Tuple& loc0, Tuple& new_t)
+// a valid mesh can have triangles that are is_removed == true
+// it can be easier for compact later on ?
+bool wmtk::TriMesh::check_mesh_connectivity_validity() const
 {
-    // TODO: use the get_next_empty_slot_v
+    std::vector<std::vector<size_t>> conn_tris(m_vertex_connectivity.size());
+    for (size_t i = 0; i < m_tri_connectivity.size(); i++) {
+        if (m_tri_connectivity[i].m_is_removed) continue;
+        for (int j = 0; j < 3; j++) conn_tris[m_tri_connectivity[i][j]].push_back(i);
+    }
+
+    // check conn_tets duplication, order, amount ...
+    for (size_t i = 0; i < m_vertex_connectivity.size(); i++) {
+        if (m_vertex_connectivity[i].m_is_removed) continue;
+        assert(
+            m_vertex_connectivity[i].m_conn_tris == conn_tris[i] &&
+            "m_vertex_connectivity[i].m_conn_tris!=conn_tris[i]");
+    }
+    return true;
+}
+
+// link check, prerequisite for edge collapse
+bool wmtk::TriMesh::check_link_condition(const Tuple& edge) const
```

Comment:
link condition should have its own test.

Reply:
There is one

Reply:
Thanks, I see it now.